### ISSUE TITLE
History-matching and MF6?

### ISSUE THREAD

**Issue by nikobenho** [pypest/pyemu#247]
Hi,

I suspect this is a stupid question with a simple answer, but I have failed to get it working these last two days so here goes:

I am developing a MF6 model. To get it set up with the PEST interface, I have followed the example outlined in the [pyemu_pestpp_workflow](https://github.com/pypest/pyemu_pestpp_workflow) repo, using PstFrom. Thus far, the model is parameterized as expected and I've been able to get prior MC going.

However, following the example above, in order to add observations to the pest control-file:
```
hds_df = pf.add_observations(
    "heads.csv",  # model output file to use 
    insfile="heads.csv.ins",  # optional, define name of PEST instruction file
    index_cols="time",  # column used to index observation/outputs
    use_cols=list(df.columns.values),  # columns to setup observations for (can be multiple)
    prefix="hds",  # observation name prefix
)
```
I find that the observations being added to the control-file are those of the simulated model outputs, and not the in-field measured heads. I suspect this to be the reason why pst.phi is really small (0.007). Basically, PEST thinks the measured heads are more or less equal to what the model is already outputting.

Trying to understand what was going on and how to properly assign measured heads, I found this similar [issue](https://github.com/pypest/pyemu/issues/109), as well as examples in the two repos: [GW1876](https://github.com/mnfienen/GW1876) and [decision_support_analysis_notebooks](https://github.com/jtwhite79/decision_support_analysis_notebooks)
However, these examples use the `pyemu.helpers.PstFromFlopyModel` class which is tailored for MF2005 models. Trying to follow the examples anyway, using `kij_dict` and `frun_line,tr_hds_df = pyemu.gw_utils.setup_hds_timeseries(binary_file,kij_dict=kij_dict,include_path=True,model=m)` lead to `AttributeError: 'ModflowGwfdis' object has no attribute 'itmuni'`, which makes me assume that `pyemu.gw_utils.setup_hds_timeseries` is designed to be used with MF2005 models specifically? At any rate, since MF6 through its OBS package is designed to output specified model outputs itself (as shown in the [pyemu_pestpp_workflow](https://github.com/pypest/pyemu_pestpp_workflow) repo), the model simulated output part of the history-matching procedure is already "handled" (i've identified which cell-ids correspond to which observations and made Flopy/MF6 output these after each time step).

So, basically, what I have working right now is:
1. measured heads
2. model simulated counterparts (as outputted to .csv by the mf6 OBS package)

And what I want to do is:
1. input measured heads as observations into the PEST control file (rather than model simulated heads)
2. have PEST++ be able to compare measured heads and simulated heads after each run in the forward_run.py (similar to the two MF2k5 examples above that use `pyemu.gw_utils.setup_hds_timeseries`, but for a MF6 environment).

As with most scenarios that involve me with modeling and programming (alot lately), I just assume I'm an idiot and the solution is right infront of me, although I have a hard time finding it. :laughing:

Any help is much appreciated!

**Comment by jtwhite79** [comment]
The simulated values are put in the control file on purpose.  That's so you can check that the pest interface is working as expected (very low phi) and also because setting the observed counterpart to the simulated output is probably the hardest thing about using pest(++) and also one of the most important things.  So we leave that to the user to align/process the simulated outputs and corresponding observations, and then set the obsvals and weights in the control file.  If you are comfortable with pandas, then you can do this via the 'Pst.observation_data' dataframe after you have called 'PstFrom.build_pst()'.  You can also save the control file as "version 2" which will save the observation data as a csv for manipulation in excel if that's easier for you.

I think the mf6 obs packages eliminates the need for that hds timeseries util since you can get mf6 to extract all simulated heads from a given node and write them to a csv ( essentially what that util does too).

**Comment by nikobenho** [comment]
@jtwhite79 Thanks, got it working! Super intuitive actually now that I understand it (which confirms my previous hypothesis).

While on the topic, perhaps I can ask if there's a suggested way of setting up temporal-differences as observations? And secondary-model outcomes (as described in the OLPROC documentation), e.g. head at different depths in a nested piezometer, or head/stage differences, but using Flopy/PyEMU/Pandas?

Many thanks!

**Comment by jtwhite79** [comment]
There's not anything pre-baked for that.  But if you are using `PstFrom`, then you can write a stand alone python function that does the differencing (however you like: spatial, temporal, combinations, etc) and writes a csv file.  Then you call `PstFrom.add_observations()` for that csv file and add the py function to be called after the model runs with `PstFrom.add_py_function()` - that call will search thru a source python file, looking for the function name.  It then copies the function exactly into the top of the forward run script and you tell `PstFrom` if that function should be called before or after the model runs.  If your function uses pandas, then dont forget to `PstFrom.extra_py_imports.append("pandas")`.  The trick is in handling the paths bit - that function should be designed to run in the same dir as the model.  There is an example using this function trickery in the pstfrom mf6 demo notebook:[https://github.com/pypest/pyemu/blob/develop/examples/pstfrom_mf6.ipynb](https://github.com/pypest/pyemu/blob/develop/examples/pstfrom_mf6.ipynb)

**Comment by nikobenho** [comment]
Thanks a lot @jtwhite79! I think you are referring to the `pf.add_py_function("helpers.py","process_model_outputs()",is_pre_cmd=False)` part of the example notebook, right? I will give it a try. Closing this now since I got the answer I needed for the question in the title. :) Cheers!
